### ISSUE TITLE
Sankey nodeAlign?

### ISSUE THREAD

**Issue by dangerismycat** [plotly/plotly.js#1938]
Curious if there are any plans to merge the more recent `d3-sankey` support for node alignment (from [this commit](https://github.com/d3/d3-sankey/commit/e4475b1b4f0c4409d052d7e82088f0fff963c184), June 13 of this year) into your fork? It's a really useful feature that I'd love to be able to use with your great wrapper.

**Comment by etpinard** [comment]
cc @monfera

**Comment by monfera** [comment]
@etpinard please tell me when to schedule it:

Some of our PRs to upstream `d3-sankey` were merged by Mike Bostock, so catching up makes sense for multiple reasons. One of our PRs predictably wasn't taken up, we'll need to reapply, likely not a lot of divergence yet. The alignment options should be exposed via `plotly.js` attributes in the JSON.

**Comment by etpinard** [comment]
> please tell me when to schedule it:

Definitively not a high priority for now, unless @jackparmer thinks otherwise.

**Comment by jackparmer** [comment]
Yeah, no customers on my radar who are asking about this. Probably shelve for now.

**Comment by phillyo** [comment]
would be great to have this option! when visualizing process flows, it makes no sense that the terminal nodes are all on the right of the viz (as for those which did not pass the whole process, they should not be on the same level as all the nodes which did).

**Comment by antoinerg** [comment]
@harisbal How much control do you want to have over the position of the nodes. Would it be convenient if you could specify the `x` and `y` position of each node?

**Comment by harisbal** [comment]
@antoinerg controlling doesn't sound like a bad idea indeed. What I was thinking though is being able to control the number of "vertical" groups if that makes sense. For instance in the following example 
![sankey_nodes](https://user-images.githubusercontent.com/6556306/51969798-fb787c00-247d-11e9-9e10-94775f0ae9f0.PNG)
I cannot place the "International aviation" node at the same vertical level as "Electricity grid" and "District heating" (in the picture I'm dragging it manually)
Ideally I would like to control how to group nodes (e.g. with a dictionary).
In the very nice implementation of ricklupton [https://nbviewer.jupyter.org/github/ricklupton/ipysankeywidget/blob/master/examples/More%20examples.ipynb](url) this is called rank sets
Thank you very much for checking on this.

**Comment by JeroenSchmidt** [comment]
@harisbal thank you for the link showing how to use ranksets. This is a really handy feature when one wants to order their sankey data according to a temporal layout - for example when plotting the evolution of communities that appear and disappear over time.

**Comment by harisbal** [comment]
@JeroenSchmidt that;s exactly the issue I'm facing, where I try to depict some temporal data (often disappearing between nodes).

**Comment by antoinerg** [comment]
@JeroenSchmidt @harisbal Thank you for reminding me of the need for this worthwhile feature.

In the meantime, PR https://github.com/plotly/plotly.js/pull/3583 (set to be released in `1.46.0`) allows one to manually control the position of nodes either via dragging or by setting attributes `node.(x|y)`. I am investigating providing control over the number of colums and ordering within them.

**Comment by gabubellon** [comment]
I endorse how interesting this feature would be.
I have work with d3 / d3-sankey "vanilla" and had problems with depth that in Plotly were solved.
However having a nodeAlign implementation for time series would help a lot.
Mainly to align the roots (eith no target links) more effectively.

**Comment by cordialgerm** [comment]
@antoinerg are there any updates on plans for specifying number of columns and ordering of nodes within those columns?

**Comment by ryankinsey1** [comment]
I'm interested in this as well. What I would like to be able to is have 'terminal nodes' in the middle (i.e. some data flows end in the middle while other's flow through to the end and/or break down into further subgroups.)

**Comment by piccinnigius** [comment]
@antoinerg any update or trick to implement this? i want to put the terminal nodes by levels (aligned to left instead to right) 

this: 
![image](https://user-images.githubusercontent.com/10252011/69914763-3bb2dd00-1448-11ea-9068-1b5270467021.png)

instead of this:
![image](https://user-images.githubusercontent.com/10252011/69914756-1f16a500-1448-11ea-98c5-2096057ad0e8.png)

it's bad to show this kind of things in production...
oooooooor how can export the full sankey data with fixed positions?

thanks so much for help!

**Comment by antoinerg** [comment]
This feature should not be hard to implement since it is supported in our dependencies. I started a branch a while ago https://github.com/plotly/plotly.js/compare/sankey2-node-align that shows what needs to be done. After resolving the conflicts, it might be good to go. I don't have the time to work on this at the moment but I can revisit later of help someone make a PR.

**Comment by tanvi-cota** [comment]
Do we have any update on this feature? Are we looking to work on this soon? This is a good feature, and already available in our dependency.

**Comment by nicolaskruchten** [comment]
No one from the Plotly team is working on this at the moment but we’d be happy to accept a PR if someone wants to take Antoine’s work over the finish line :)

**Comment by tanvi-cota** [comment]
@nicolaskruchten I would love to contribute, let me see how can i get that updated.

**Comment by jackparmer** [comment]
This issue has been tagged with `NEEDS SPON$OR`

A [community PR](https://github.com/plotly/plotly.js/pulls?q=is%3Apr+is%3Aopen+label%3A%22type%3A+community%22) for this feature would certainly be welcome, but our experience is deeper features like this are difficult to complete without the Plotly maintainers leading the effort.

**Sponsorship range: $5k-$10k**

What Sponsorship includes:
- Completion of this feature to the Sponsor's satisfaction, in a manner coherent with the rest of the Plotly.js library and API
- Tests for this feature
- Long-term support (continued support of this feature in the latest version of Plotly.js)
- Documentation at [plotly.com/javascript](https://plotly.com/javascript/)
- Possibility of integrating this feature with Plotly Graphing Libraries (Python, R, F#, Julia, MATLAB, etc)
- Possibility of integrating this feature with Dash
- Feature announcement on [community.plotly.com](https://community.plotly.com/tag/announcements) with shout out to Sponsor (or can remain anonymous)
- Gratification of advancing the world's most downloaded, interactive scientific graphing libraries (>50M downloads across supported languages)

Please include the link to this issue when [contacting us](https://plotly.com/get-pricing/) to discuss.

**Comment by KaranTrivedi** [comment]
Bump on this issue? \

**Comment by gaetpinon** [comment]
Hi everyone, I saw an 'align' property was added in plotly v5.19.0 to solve this issue. Unfortunately I can't make it work. I can try any of the 4 values ['center', 'right', 'left', 'justify'], but nothing will change. I tried with the same example as in the [plotly sankey tutorial](https://plotly.com/python/sankey-diagram/#node-alignment). I have the latest plotly version v5.19.0

Thank you for your help
@adamreeve @archmoj
